### PR TITLE
Undo message correctly states what is going to be undone

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
David made a pretty good remark regarding a PR I made: https://github.com/ankidroid/Anki-Android/pull/6863#discussion_r467661553 

And he was extremely right. Actually nothing is safe regarding the queue of undoables. Background tasks can add things any time. The queue could theoretically be emptied (seems hard to do in practice). 

Each time `undo` was called, it was undoing the last element from the queue. Which could seems to be a good idea except that this element may not exists. Or it may have changed, and so the actual undone action would differ from the one you expected to undo. This would be extremely nonsensical in card browser where it asks you whether you want to undelete a bunch a note and you risk to undo something else by accident if you wait long enough.

I'm going to test this on my phone. 

I must state I'm quite happy to have added Undoable. If data were still an array of object, I fear it would have been harder to follow through the logic of this correction

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/libanki/Undoable.java [hunk 1]
```diff
@@ -33,18 +35,33 @@ public DismissType getDismissType() {
...
 
+
     /**
      * Return MULTI_CARD when no other action is needed, e.g. for multi card action
      * Return NO_REVIEW when we just need to reset the collection
+     * Return NOT_UNDONE if the undo was already started and so is not executed here
      * Returned positive integers are card id. Those ids is the card that was discarded and that may be sent back to the reviewer.*/
-    public abstract long undo(Collection col);
+    public long undo(Collection col){
+        synchronized (mStarted) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Undoable.java:45]: Don't lock on a mutable object: https://stackoverflow.com/questions/10324272/why-is-it-not-a-good-practice-to-synchronize-on-boolean

File: AnkiDroid/src/main/java/com/ichi2/libanki/Undoable.java [hunk 2]
```diff
@@ -33,18 +35,33 @@ public DismissType getDismissType() {
...
+            if (mStarted) {
+                return NOT_UNDONE;
+            }
+            mStarted = true;
+        }
+        return actualUndo(col);
+    }
+    /**Actually undo. It assumes the action was not already undone. Return same as undo, but can't return NOT_UNDONE */
+    protected abstract long actualUndo(Collection col);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Undoable.java:54]: (Optional) I think this feels slightly more descriptive. 
```suggestion
    protected abstract long performUndoOperation(Collection col);
```

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java [hunk 3]
```diff
@@ -1294,9 +1311,12 @@ public long undo() {
     public void markUndo(Undoable undo) {
         Timber.d("markUndo() of type %s", undo.getDismissType());
         LinkedList<Undoable> undos = mUndo; // Copied for synchronization
-        undos.add(undo);
-        while (undos.size() > UNDO_SIZE_MAX) {
-            undos.removeFirst();
+        synchronized(undos) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:1314]: Locking on a local variable here is strange, what are you trying to accomplish here?

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java [hunk 4]
```diff
@@ -1250,10 +1250,11 @@ public void clearUndo() {
     }
 
     public @Nullable Undoable lastUndo() {
-        if (mUndo.isEmpty()) {
+        LinkedList<Undoable> undoables = mUndo; // Copied for synchronization
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:1256]: Only useful if the reference is going to change, not the items

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java [hunk 5]
```diff
@@ -1288,11 +1289,13 @@ public long undo() {
         return NO_REVIEW;
     }
 
+    /* Undos may get cleared during execution of markUndo. In this case the action can't be undone*/
     public void markUndo(Undoable undo) {
         Timber.d("markUndo() of type %s", undo.getDismissType());
-        mUndo.add(undo);
-        while (mUndo.size() > UNDO_SIZE_MAX) {
-            mUndo.removeFirst();
+        LinkedList<Undoable> undos = mUndo; // Copied for synchronization
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:1313]: ditto

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java [hunk 6]
```diff
@@ -1251,13 +1252,29 @@ public void clearUndo() {
...
            return null;
         }
         return undoables.getLast();
     }
 
+    /**
+     * [type, undoName, data] type 1 = review; type 2 =
+     */
+    public void cleanUndo() {
+        synchronized (mUndo) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:1267]: This either needs a cleaner implementation, or better naming/commenting, I don't see:

* When this should be called
* Why it exists

Why aren't tasks just removed from the list when they're started?

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java [hunk 7]
```diff
@@ -359,6 +360,9 @@ protected TaskData actualDoInBackground(TaskData param) {
             case UNDO:
                 return doInBackgroundUndo();
 
+            case UNDO_NAMED:
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:363]: This needs a better name, or maybe UNDO does. I don't have good suggestions right now.

UNDO_PROVIDED_VALUE maybe?

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java [hunk 8]
```diff
@@ -1166,6 +1170,53 @@ private TaskData doInBackgroundUndo() {
     }
 
 
+    private TaskData doInBackgroundUndoNamed(TaskData param) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:1170]: Is this what we want? Breaking the guarantee that mUndo is a stack is scary and hard to verify.

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 9]
```diff
@@ -183,6 +184,8 @@
 
     private TextView mReviewSummaryTextView;
 
+    private Undoable mUndoable;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:187]:
```suggestion
    @Nullable
    private Undoable mUndoable;
```

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 10]
```diff
@@ -1273,8 +1277,7 @@ private void upgradePreferences(long previousVersionCode) {
 
     private void undo() {
         Timber.i("undo()");
-        String undoReviewString = getResources().getString(R.string.undo_action_review);
-        final boolean isReview = undoReviewString.equals(getCol().undoName(getResources()));
+        final boolean isReview = undoReviewString.equals(mUndoable.Name(getResources()));
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java]: `Name` doesn't match the Java conventions

File: AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.java [hunk 11]
```diff
@@ -295,7 +298,7 @@ public boolean onMenuItemClick(MenuItem item) {
         switch (item.getItemId()) {
             case R.id.action_undo:
                 Timber.i("StudyOptionsFragment:: Undo button pressed");
-                getCol().undo();
+                mUndoable.undo(getCol());
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/StudyOptionsFragment.java:301]: Can Ctrl+Z trigger this? Maybe we want more of a guard than visibility

File: AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java [hunk 12]
```diff
@@ -1103,6 +1095,11 @@ private TaskData doInBackgroundDismissNotes(TaskData param) {
             AnkiDroidApp.sendExceptionReport(e, "doInBackgroundSuspendCard");
             return new TaskData(false);
         }
+
+        if (undoable != null) {
+            // mark undo for all at once
+            col.markUndo(undoable);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/async/CollectionTask.java:1101]: This is now performed outside the transaction

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java [hunk 13]
```diff
@@ -1271,9 +1280,12 @@ public boolean undoAvailable() {
     }
 
     public long undo() {
-        Undoable lastUndo = mUndo.removeLast();
+        Undoable lastUndo = lastUndo();
```
[the remark this answers is not in the document]
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:1302]: This should still be NF: if an element is received, then `lastUndo.undo()` mark it as started. Then `lastUndo` will clean it from the list at next call and so it won't be returned again

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java (hunk 3 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:1314]: Ensuring that while some code add a member to the list (and maybe remove some other members), no other code change the list. Indeed, if that and cleaning were running simultaneously, the loop in cleaning would break

File: AnkiDroid/src/main/java/com/ichi2/libanki/Undoable.java (hunk 1 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Undoable.java:45]: Thanks very much. I fear I still have a really bad idea of what locking actually is :( 
I'll change it if the boolean remain. If we only cancel head of the queue it will become useless

File: AnkiDroid/src/main/java/com/ichi2/libanki/Undoable.java (hunk 2 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Undoable.java:54]: I'll change it if the boolean remain. If we only cancel head of the queue, then I won't need to add an extra member

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java (hunk 10 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java]: Indeed. I wonder how it was in the PR, it does not even seen to be in my code right now. I didn't even define a method `String Name`...

**Comment by mikehardy** [comment]
These are great questions @david-allison-1 

> I like the idea of thread safety, but I think this needs a bit of a design discussion @mikehardy
> 
>     * If an undo is performed, and another undoable task has been executed, what should happen?
>       
>       * I'd personally notify the user that they should try again, rather than breaking the fact that undo is a stack.
>       * We also have the chance to pop the stack to the provided point

I like this idea best, it does what the user asked, possibly undoing a bit more than they asked, but definitely to the point they asked for, and should be provably safe every time

>       * We currently undo the provided task, which feels dangerous, but could be fine

On reflection this does seem dangerous. You can't take a stack of actions that are possibly linked via data-dependence and do anything other than pop safely.

> I think that some of these changes would be better in their own data structure class, rather than in `Collection`. I feel it's harder to reason about what's happening inside `col` after this change. Obviously I'd rather that the code were correct, but we should aim to isolate this complexity a little better

Perhaps but I'm wary of reading great chunks of code...

> This is complex, the requirement: "When a named undo is processed and there's more on the stack, then X should happen" would be a very high-value unit test.

THIS THIS THIS please a test that exercises it

**Comment by Arthur-Milchior** [comment]
> These are great questions @david-allison-1
> 
> > I like the idea of thread safety, but I think this needs a bit of a design discussion @mikehardy
> > ```
> > * If an undo is performed, and another undoable task has been executed, what should happen?
> >   
> >   * I'd personally notify the user that they should try again, rather than breaking the fact that undo is a stack.
> >   * We also have the chance to pop the stack to the provided point
> > ```
> 
> I like this idea best, it does what the user asked, possibly undoing a bit more than they asked, but definitely to the point they asked for, and should be provably safe every time

I like this idea as a user. As a dev' I don't see how to do it. Indeed, while the undo message may state "Undo X", maybe "X" was removed from the undo stack. Either because 20 actions have been made since X. Or because the list has been emptied. In those case, I can undo the entire content of the queue without actually undoing X. 

Of course, I could just check whether X belongs to the queue and if it is not the case warn the user undoing was impossible, or something similar. The trouble is that even if X is in the queue, it does not help a lot. Because if we empty the undo stack in case of memory pressure, then X can be removed from the stack while we are emptying the stack. Or maybe some Content Provider or some background task who ended are adding element in the queue while we empty it. Should this element also be deleted ? 

Of course, all of those are theoretical problem. I have not heard this problem occurring in real life. But if we want to really be thread safe, we should consider all of those cases.

This is why I suggested the far simpler solutions:
* undoing what we told the user we would undo (but I understand that it may also be risky)
* telling the user that actually, we can't because other actions were done since, so please retry (and actually we don't guarantee at all that you'll actually be able to undo anything)

> 
> > ```
> >   * We currently undo the provided task, which feels dangerous, but could be fine
> > ```
> 
> On reflection this does seem dangerous. You can't take a stack of actions that are possibly linked via data-dependence and do anything other than pop safely.

What do you mean by "provided task" ? Currently, we undo the task at the end of the stack(even if we tell the user we do something else). The provided task is undone in my PR only, but my PR is not what "currently" is done. 


> 
> > I think that some of these changes would be better in their own data structure class, rather than in `Collection`. I feel it's harder to reason about what's happening inside `col` after this change. Obviously I'd rather that the code were correct, but we should aim to isolate this complexity a little better
> 
> Perhaps but I'm wary of reading great chunks of code...

May I suggest using a tool for diffing code ? If the code move from one file to another but there is no change, it's hard to follow.

But I know this is a place where we strongly disagree. I believe it'd be far easier to read the new version of the code, where all code logically related belong together, while you prefer to read change in a lot of file/in a lot of place in a file if it allows to avoid moving everyhting.

> 
> > This is complex, the requirement: "When a named undo is processed and there's more on the stack, then X should happen" would be a very high-value unit test.
> 
> THIS THIS THIS please a test that exercises it

I don't understand. How can we do a unit test if we have not yet decided what should occur ? Once we decide, I believe I see how to create the test. I could create a test for this PR (but I won't do it since it seems you expect to reject it, for good reason)

**Comment by Arthur-Milchior** [comment]
Not goign to resolve conflict until we decide what behavior we want. In particular, what do we do if the undone operation is not in the list anymore.

**Comment by Arthur-Milchior** [comment]
@mikehardy and @david-allison-1 I am still waiting for answers to the remark I made above to know what to implements. Because I noted problem with the last proposed behavior and don't know how to deal with it in a thread-safe way, especially in an interactive environment

**Comment by mikehardy** [comment]
Sure! Thanks for your patience - you're obviously on Discord and we're chatting so you know what we've been up to - but for future so you know, I am planning on getting back to normal PR reviews and merges probably next week once the public has a chance to crash 2.13.0 a little bit and we get 2.13.1 out

**Comment by Arthur-Milchior** [comment]
What confused me here is that I was not asking for a PR review. I was asking for what behavior we would expect in a case not covered by the discussion above. So no code reading, since anyway the current version of the code has already been rejected. Only deciding what to do if the undo list has been emptied before the action mentioned in the menu had been undone.

**Comment by david-allison** [comment]
Superseded by 0407c0291f3cb796eb68e544b79026ead8b2adba (#10454)
